### DISCUSSION TITLE
RTOSでROSを動かす

### DISCUSSION THREAD

**Discussion by TadashiSakamoto** [KyotoMicrocomputer/solid-rapi4#6] [Q&A]
SLIDをインストールして、Lチカまではできています。またLinux側でのROS2(humble)は動いていますが、RTOS側でmicro-ROSまたはmROS2を動かして、リアルタイム性能を上げたいと思っています。
micro-ROSあるいはｍROS2を動かしている、またはこうすれば動くよという情報をいただけると助かります。
よろしくお願いいたします。

**Comment by kawadakk** [comment]
[mROS 2](https://github.com/mROS-base/mros2)の当環境における動作実績はまだ無いものと思われますが、[mROS-base/mros2-asp3-f767zi](https://github.com/mROS-base/mros2-asp3-f767zi) 等のサンプルコードを基に必要な部分を実装すれば動作させられる可能性があります。

ざっと見たところでは少なくとも以下の移植作業が必要になると思われます。

- **LwIPをSOLID Socketsで置き換える。**

    - embeddedRTPSライブラリの [`src/communication/UdpDriver.cpp`](https://github.com/mROS-base/embeddedRTPS/blob/de70e01959afdf218af8a227f054226f78806cdd/src/communication/UdpDriver.cpp) でLwIPが使用されています。これを[SOLID Sockets](https://solid.kmckk.com/SOLID/doc/3.2.0/os/sockets/index.html#api)の上で再実装することが必要です。

    - POSIX向けのサンプルコード [mROS-base/mros2-posix](https://github.com/mROS-base/mros2-posix) では [lwip-posix](https://github.com/mROS-base/lwip-posix/tree/f1d576bde977f20a452654ab283fa11c03a2ffc1/src/core) によりPOSIX sockets上でLwIPのAPIをエミュレートしています。これを若干修正すればSOLID Sockets [POSIX互換ラッパー](https://solid.kmckk.com/SOLID/doc/3.2.0/os/sockets/index.html#solid-sockets-compat) (`#include "solid_sockets_compat.h"`) の上で動かせるかもしれません。

    - ネットワークインタフェースの初期化はアプリケーション起動前に行われるため、必要ありません。

- **TOPPERS/FMP3カーネル (+ SOLID拡張) 向けのOS抽象化レイヤーを提供する。**

    - [mROS-base/mros2-asp3-f767zi](https://github.com/mROS-base/mros2-asp3-f767zi) では [mROS-base/cmsis-asp3](https://github.com/mROS-base/cmsis-asp3/tree/main) にあるTOPPERS/ASP3向けのCMSIS OS APIの実装を利用しています。しかし、これは静的カーネルコンフィギュレーションを前提としているため、SOLID-RPi4では使用できません。

    - ESP-IDF FreeRTOS向けのサンプルにFreeRTOS向けの実装 [`cmsis-esp32/cmsis-esp.c`](https://github.com/mROS-base/mros2-esp32/blob/b0f4b47152462fab9f313afdbe72a68f4095ab7c/cmsis-esp32/cmsis-esp.c)が含まれています。これには必要最小限の関数のみが含まれるため、これをベースにTOPPERS/FMP3向けの実装を作成するのがやりやすいのではないかと思われます。

    - 参考: [TOPPERS第3世代カーネル (ITRON系) 統合仕様書 Release 3.5.0](https://toppers.jp/docs/tech/tgki_spec-350.pdf)
    - 参考: `C:\ProgramData\KMC\SOLID\SDK\rpi4\include\rtos\toppers_fmp3\fmp3\include\kernel.h`

**Reply by TadashiSakamoto** [reply]
詳細なコメントありがとうございます。
ただ私のスキルではハードルが高すぎて、移植作業は無理と思われます。
別の構成を考えてみたいと思います。
ありがとうございました。
